recipes/3p_integrations/llamaindex/dlai_agentic_rag/Building_Agentic_RAG_with_Llamaindex_L3_Building_an_Agent_Reasoning_Loop.ipynb

```diff
@@ -0,0 +1,4264 @@
+    "    file_path: str,\n",
+    "    name: str,\n",
+    ") -> str:\n",
+    "    \"\"\"Get vector query and summary query tools from a document.\"\"\"\n",
+    "\n",
+    "    # load documents\n",
+    "    documents = SimpleDirectoryReader(input_files=[file_path]).load_data()\n",
+    "    splitter = SentenceSplitter(chunk_size=1024)\n",
+    "    nodes = splitter.get_nodes_from_documents(documents)\n",
+    "    vector_index = VectorStoreIndex(nodes)\n",
+    "\n",
+    "    def vector_query(\n",
+    "        query: str,\n",
+    "        page_numbers: Optional[List[str]] = None\n",
+    "    ) -> str:\n",
+    "        \"\"\"Use to answer questions over the MetaGPT paper.\n",
+    "\n",
+    "        Useful if you have specific questions over the MetaGPT paper.\n",
+    "        Always leave page_numbers as None UNLESS there is a specific page you want to search for.\n",
+    "\n",
+    "        Args:\n",
+    "            query (str): the string query to be embedded.\n",
+    "            page_numbers (Optional[List[str]]): Filter by set of pages. Leave as NONE\n",
+    "                if we want to perform a vector search\n",
+    "                over all pages. Otherwise, filter by the set of specified pages.\n",
+    "\n",
+    "        \"\"\"\n",
+    "\n",
+    "        page_numbers = page_numbers or []\n",
+    "        metadata_dicts = [\n",
+    "            {\"key\": \"page_label\", \"value\": p} for p in page_numbers\n",
+    "        ]\n",
+    "\n",
+    "        query_engine = vector_index.as_query_engine(\n",
+    "            similarity_top_k=2,\n",
+    "            filters=MetadataFilters.from_dicts(\n",
+    "                metadata_dicts,\n",
+    "                condition=FilterCondition.OR\n",
+    "            )\n",
+    "        )\n",
+    "        response = query_engine.query(query)\n",
+    "        return response\n",
+    "\n",
+    "\n",
+    "    vector_query_tool = FunctionTool.from_defaults(\n",
+    "        name=f\"vector_tool_{name}\",\n",
+    "        fn=vector_query\n",
+    "    )\n",
+    "\n",
+    "    summary_index = SummaryIndex(nodes)\n",
+    "    summary_query_engine = summary_index.as_query_engine(\n",
+    "        response_mode=\"tree_summarize\",\n",
+    "        use_async=True,\n",
+    "    )\n",
+    "    summary_tool = QueryEngineTool.from_defaults(\n",
+    "        name=f\"summary_tool_{name}\",\n",
+    "        query_engine=summary_query_engine,\n",
+    "        description=(\n",
+    "            \"Use ONLY IF you want to get a holistic summary of MetaGPT. \"\n",
+    "            \"Do NOT use if you have specific questions over MetaGPT.\"\n",
+    "        ),\n",
+    "    )\n",
+    "\n",
+    "    return vector_query_tool, summary_tool"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "6zi9eRc2dOMQ"
+   },
+   "outputs": [],
+   "source": [
+    "vector_tool, summary_tool = get_doc_tools(\"metagpt.pdf\", \"metagpt\")"
```

Comment:
It looks for an API key for OpenAI when calling the `get_doc_tools`. Is there a prerequisite to set `OPENAI_API_KEY` or `openai.api_key` before running?